File: src/rust/flows_development/src/verify.rs
//! Automatic verification: runs `verify-feature.sh <id>` in the target directory, with a
//! time ceiling (derived from `harness_config.timeout_ms`) and a full log in
//! `.harness/logs/`. When `RunConfig.verify_cmds` holds a non-empty list instead, every
//! command in it runs CONCURRENTLY (one log each), with an AND verdict — see
//! `try_parallel_configured_verify`.

use std::io::Read;
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use std::thread;
use std::time::{Duration, Instant};

use harness_engine::harness_config;

pub struct AutomatedVerifyResult {
    pub attempted: bool,
    pub success: bool,
    pub result: String,
}

impl AutomatedVerifyResult {
    fn missing() -> Self {
        Self {
            attempted: false,
            success: false,
            result: String::new(),
        }
    }

    fn passed(result: String) -> Self {
        Self {
            attempted: true,
            success: true,
            result,
        }
    }

    fn failed(result: String) -> Self {
        Self {
            attempted: true,
            success: false,
            result,
        }
    }
}

struct VerifyScriptResult {
    exit_code: i32,
    output: String,
    error: String,
    timed_out: bool,
}

pub fn try_automated_verify(feature_id: i32, target_dir: &Path, verify_cmd: &str, verify_cmds: &[String]) -> AutomatedVerifyResult {
    let script = target_dir.join("verify-feature.sh");
    let (command, label, is_script) = if script.is_file() {
        (vec!["bash".to_string(), script.to_string_lossy().to_string(), feature_id.to_string()], format!("bash ./verify-feature.sh {feature_id}"), true)
    } else if !verify_cmds.is_empty() {
        // Configured as a list (RunConfig.verify_cmds): run every command concurrently,
        // AND verdict, instead of the single-command path below. Takes precedence over
        // `verify_cmd` but not over `verify-feature.sh`, same ordering as the single
        // command had relative to the script.
        return try_parallel_configured_verify(target_dir, verify_cmds, feature_id);
    } else {
        let args = configured_verify_argv(verify_cmd);
        if args.is_empty() { return AutomatedVerifyResult::missing(); }
        (args.clone(), args.join(" "), false)
    };

    let result = run_verify_script(target_dir, &command);
    let log_path = write_verify_log(target_dir, &label, feature_id, &result, None);

    if result.timed_out {
        return AutomatedVerifyResult::failed(format!(
            "FAIL: verification exceeded timeout ({}){}",
            verify_timeout_description(),
            verify_output_suffix(&result, &log_path)
        ));
    }

    if result.exit_code == 0 {
        return AutomatedVerifyResult::passed(format!("PASS: {} passed{}", if is_script { format!("verify-feature.sh {feature_id}") } else { "configured verify command".to_string() }, log_suffix(&log_path)));
    }

    AutomatedVerifyResult::failed(format!(
        "FAIL: verification failed (exit {}){}",
        result.exit_code,
        verify_output_suffix(&result, &log_path)
    ))
}

/// Runs every entry in `commands` CONCURRENTLY (one `std::thread::spawn` each — no
/// tokio/rayon in this workspace, mirroring the reader-thread pattern in
/// `run_verify_script` and the abandon-on-timeout thread in
/// `task_registry::run_with_timeout`), and ANDs the verdicts: passes only if every command
/// exits 0. Each command gets its own log (`verify-feature-{id}-{n}.log`, 1-based `n`), and
/// the aggregate message lists every command's outcome, not just the first failure.
fn try_parallel_configured_verify(target_dir: &Path, commands: &[String], feature_id: i32) -> AutomatedVerifyResult {
    // Tokenize everything up front, before launching any subprocess: a single disallowed
    // command fails the whole check immediately, without side-launching the others.
    let mut argvs: Vec<Vec<String>> = Vec::with_capacity(commands.len());
    for (i, raw) in commands.iter().enumerate() {
        let args = configured_verify_argv(raw);
        if args.is_empty() {
            return AutomatedVerifyResult::failed(format!(
                "FAIL: verify command #{} is empty or uses disallowed shell operators: {raw}",
                i + 1
            ));
        }
        argvs.push(args);
    }

    let target_dir_owned = target_dir.to_path_buf();
    let handles: Vec<_> = argvs
        .into_iter()
        .map(|args| {
            let target_dir = target_dir_owned.clone();
            thread::spawn(move || {
                let label = args.join(" ");
                let result = run_verify_script(&target_dir, &args);
                (label, result)
            })
        })
        .collect();

    let total = handles.len();
    let mut failures = 0usize;
    let mut lines: Vec<String> = Vec::with_capacity(total);
    for (i, handle) in handles.into_iter().enumerate() {
        let n = i + 1;
        match handle.join() {
            Ok((label, result)) => {
                let log_path = write_verify_log(target_dir, &label, feature_id, &result, Some(n));
                if result.timed_out {
                    failures += 1;
                    lines.push(format!(
                        "#{n} TIMEOUT ({}): {label}{}",
                        verify_timeout_description(),
                        log_suffix(&log_path)
                    ));
                } else if result.exit_code == 0 {
                    lines.push(format!("#{n} PASS: {label}{}", log_suffix(&log_path)));
                } else {
                    failures += 1;
                    lines.push(format!(
                        "#{n} FAIL (exit {}): {label}{}",
                        result.exit_code,
                        verify_output_suffix(&result, &log_path)
                    ));
                }
            }
            // A worker thread panicking (a bug, not a check failure) must not take down the
            // whole verify step — treated as a failed check for this one command, the
            // others' results are unaffected since each thread is independent.
            Err(_) => {
                failures += 1;
                lines.push(format!(
                    "#{n} FAIL: verify command panicked before completing: {}",
                    commands[i]
                ));
            }
        }
    }

    let joined = lines.join(" | ");
    if failures == 0 {
        AutomatedVerifyResult::passed(format!("PASS: all {total} verify commands passed. {joined}"))
    } else {
        AutomatedVerifyResult::failed(format!(
            "FAIL: {failures} of {total} verify commands did not pass. {joined}"
        ))
    }
}

fn run_verify_script(target_dir: &Path, command: &[String]) -> VerifyScriptResult {
    let child = Command::new(&command[0])
        .args(&command[1..])
        .current_dir(target_dir)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn();

    let mut child = match child {
        Ok(c) => c,
        Err(e) => {
            return VerifyScriptResult {
                exit_code: -1,
                output: String::new(),
                error: e.to_string(),
                timed_out: false,
            };
        }
    };

    // Reader threads drain the pipes continuously — without this, a script with large
    // output would hang writing into the full pipe while the poll loop below only
    // observes the status, without reading anything (the same problem that .NET's
    // `ReadToEndAsync` and Python's `subprocess.run(timeout=...)` internal drain avoid).
    let mut stdout_pipe = child.stdout.take().expect("stdout piped");
    let mut stderr_pipe = child.stderr.take().expect("stderr piped");
    let stdout_handle = thread::spawn(move || {
        let mut buf = String::new();
        let _ = stdout_pipe.read_to_string(&mut buf);
        buf
    });
    let stderr_handle = thread::spawn(move || {
        let mut buf = String::new();
        let _ = stderr_pipe.read_to_string(&mut buf);
        buf
    });

    let timeout_ms = verify_timeout_ms();
    let deadline = if timeout_ms > 0 {
        Some(Instant::now() + Duration::from_millis(timeout_ms as u64))
    } else {
        None
    };

    let (exit_code, timed_out) = loop {
        match child.try_wait() {
            Ok(Some(status)) => break (status.code().unwrap_or(-1), false),
            Ok(None) => {
                if let Some(deadline) = deadline {
                    if Instant::now() >= deadline {
                        let _ = child.kill();
                        let _ = child.wait();
                        break (-1, true);
                    }
                }
                thread::sleep(Duration::from_millis(20));
            }
            Err(_) => break (-1, false),
        }
    };

    VerifyScriptResult {
        exit_code,
        output: stdout_handle.join().unwrap_or_default(),
        error: stderr_handle.join().unwrap_or_default(),
        timed_out,
    }
}

// Verify's time ceiling: a margin under the global timeout, so the harness still has a
// chance to report the overrun before dispatch's own step guard cuts it off.
fn verify_timeout_ms() -> i32 {
    let timeout_ms = harness_config::current().timeout_ms;
    if timeout_ms <= 0 {
        return 0;
    }
    let margin = (timeout_ms / 10).clamp(1, 500);
    (timeout_ms - margin).max(1)
}

fn verify_timeout_description() -> String {
    let timeout_ms = verify_timeout_ms();
    if timeout_ms <= 0 {
        "no limit".to_string()
    } else {
        format!("{timeout_ms}ms")
    }
}

// `check_index` distinguishes one of several concurrent verify commands (1-based, from
// `try_parallel_configured_verify`) from the single-command path: `Some(n)` names the log
// `verify-feature-{id}-{n}.log`, `None` keeps the original `verify-feature-{id}.log`.
fn write_verify_log(
    target_dir: &Path,
    command: &str,
    feature_id: i32,
    result: &VerifyScriptResult,
    check_index: Option<usize>,
) -> String {
    let relative_dir = ".harness/logs";
    let file_name = match check_index {
        Some(n) => format!("verify-feature-{feature_id}-{n}.log"),
        None => format!("verify-feature-{feature_id}.log"),
    };
    let relative_path: PathBuf = [relative_dir, &file_name].iter().collect();
    let display_path = relative_path.to_string_lossy().replace('\\', "/");

    let full_path = relative_path.clone();
    let write_result = full_path
        .parent()
        .map(std::fs::create_dir_all)
        .unwrap_or(Ok(()))
        .and_then(|_| {
            std::fs::write(
                &full_path,
                format!(
                    "timestampUtc: {}\n\
                    command: {command}\n\
cwd: {}\n\
exitCode: {}\n\
timedOut: {}\n\
\n\
--- stdout ---\n\
{}\n\
\n\
--- stderr ---\n\
{}",
                    chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Secs, true),
                    target_dir.display(),
                    result.exit_code,
                    result.timed_out,
                    result.output,
                    result.error
                ),
            )
        });

    match write_result {
        Ok(()) => display_path,
        Err(e) => format!(
            "log unavailable ({})",
            crate::handoff::one_line(&e.to_string(), "")
        ),
    }
}

fn configured_verify_argv(raw: &str) -> Vec<String> {
    let text = raw.trim();
    if text.is_empty() || [";", "&", "|", "<", ">", "`", "$"].iter().any(|x| text.contains(x)) { return Vec::new(); }
    let args = text.split_whitespace().map(str::to_string).collect::<Vec<_>>();
    if args.is_empty() { return args; }
    let bin = std::path::Path::new(&args[0]).file_name().and_then(|x| x.to_str()).unwrap_or("").to_ascii_lowercase();
    if ["sh", "bash", "zsh", "fish", "cmd", "powershell", "pwsh"].contains(&bin.as_str()) && args.iter().skip(1).any(|x| ["-c", "-command", "/c"].contains(&x.as_str())) { return Vec::new(); }
    args
}

fn verify_output_suffix(result: &VerifyScriptResult, log_path: &str) -> String {
    let output = snippet(&first_meaningful_line(&[&result.output, &result.error]));
    if output.trim().is_empty() {
        log_suffix(log_path)
    } else {
        format!(": {output}{}", log_suffix(log_path))
    }
}

fn first_meaningful_line(values: &[&str]) -> String {
    for value in values {
        for line in value.replace('\r', "\n").split('\n') {
            let trimmed = line.trim();
            if !trimmed.is_empty() {
                return trimmed.to_string();
            }
        }
    }
    String::new()
}

fn log_suffix(log_path: &str) -> String {
    if log_path.trim().is_empty() {
        String::new()
    } else {
        format!(". Log: {log_path}")
    }
}

fn snippet(value: &str) -> String {
    let text = crate::handoff::one_line(value, "");
    const MAX_BYTES: usize = 240;
    if text.len() <= MAX_BYTES {
        text
    } else {
        let truncated = crate::handoff::truncate_utf8_bytes(&text, MAX_BYTES);
        truncated.trim_end().to_string() + "..."
    }
}

File: src/go/flowsdevelopment/verify.go
package main

import (
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	engine "github.com/cooperacode/IAO/src/go/harnessengine"
)

type automatedVerifyResult struct {
	Attempted bool
	Success   bool
	Result    string
}

func missingVerify() automatedVerifyResult { return automatedVerifyResult{} }
func passedVerify(result string) automatedVerifyResult {
	return automatedVerifyResult{Attempted: true, Success: true, Result: result}
}
func failedVerify(result string) automatedVerifyResult {
	return automatedVerifyResult{Attempted: true, Success: false, Result: result}
}

func tryAutomatedVerify() automatedVerifyResult {
	featureId, err := strconv.Atoi(state(currentFeatureIdKey))
	if err != nil {
		return missingVerify()
	}

	targetDir, err := resolveTargetDir(engine.LoadRunConfig().TargetDir)
	if err != nil {
		// Invalid target_dir (root, home, harness install) -> same "didn't attempt
		// automated verification" path as a target_dir with no verify-feature.sh; doesn't
		// bring the process down with an unhandled error.
		engine.LogError(fmt.Sprintf("[dev] invalid target_dir for automatic verify: %s", err))
		return missingVerify()
	}

	script := filepath.Join(targetDir, "verify-feature.sh")
	isScript := fileExistsLocal(script)

	// verify-feature.sh still wins over everything else (unchanged precedent). Below that,
	// a configured list of commands (VerifyCmds) takes priority over the single configured
	// VerifyCmd — a non-empty list is a strictly more specific choice by whoever ran `plan`.
	if !isScript {
		if commands := engine.LoadRunConfig().VerifyCmds; len(commands) > 0 {
			return tryParallelConfiguredVerify(targetDir, commands, featureId)
		}
	}

	command := []string{"bash", script, strconv.Itoa(featureId)}
	label := fmt.Sprintf("bash ./verify-feature.sh %d", featureId)
	if !isScript {
		command = configuredVerifyArgv(engine.LoadRunConfig().VerifyCmd)
		if len(command) == 0 {
			return missingVerify()
		}
		label = strings.Join(command, " ")
	}

	result := runVerifyScript(targetDir, command)
	logPath := writeVerifyLog(targetDir, label, featureId, result)
	if result.TimedOut {
		return failedVerify(fmt.Sprintf("FAIL: verification exceeded timeout (%s)%s",
			verifyTimeoutDescription(), verifyOutputSuffix(result, logPath)))
	}

	if result.ExitCode == 0 {
		kind := "configured verify command"
		if isScript {
			kind = fmt.Sprintf("verify-feature.sh %d", featureId)
		}
		return passedVerify(fmt.Sprintf("PASS: %s passed%s", kind, logSuffix(logPath)))
	}

	return failedVerify(fmt.Sprintf("FAIL: verification failed (exit %d)%s",
		result.ExitCode, verifyOutputSuffix(result, logPath)))
}

type verifyScriptResult struct {
	ExitCode int
	Output   string
	Error    string
	TimedOut bool
}

func runVerifyScript(targetDir string, command []string) verifyScriptResult {
	cmd := exec.Command(command[0], command[1:]...)
	cmd.Dir = targetDir
	// New process group so a timeout can kill the whole tree, not just the direct child.
	// Implemented per-OS in process_unix.go/process_windows.go.
	setProcessGroup(cmd)

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Start(); err != nil {
		return verifyScriptResult{ExitCode: -1, Error: err.Error()}
	}

	done := make(chan error, 1)
	go func() { done <- cmd.Wait() }()

	timeoutMs := verifyTimeoutMs()
	if timeoutMs <= 0 {
		err := <-done
		return verifyScriptResult{ExitCode: exitCodeOf(err), Output: stdout.String(), Error: stderr.String()}
	}

	select {
	case err := <-done:
		return verifyScriptResult{ExitCode: exitCodeOf(err), Output: stdout.String(), Error: stderr.String()}
	case <-time.After(time.Duration(timeoutMs) * time.Millisecond):
		// Kill the whole process group; the script may have spawned children.
		killProcessTree(cmd)
		<-done
		return verifyScriptResult{ExitCode: -1, Output: stdout.String(), Error: stderr.String(), TimedOut: true}
	}
}

// parallelVerifyOutcome is one command's contribution to the aggregate verdict computed by
// tryParallelConfiguredVerify — carried back over a channel, keyed by its original 0-based
// position so results can be reassembled in the order the commands were configured in,
// regardless of which goroutine finishes first.
type parallelVerifyOutcome struct {
	index  int
	passed bool
	line   string
}

// tryParallelConfiguredVerify runs every entry of RunConfig.VerifyCmds CONCURRENTLY and
// combines them with AND logic: the feature only passes if every command exits 0. Each
// command gets its own log file (verify-feature-{featureId}-{n}.log, 1-based) so concurrent
// runs never race on the same file — see writeVerifyLogIndexed.
//
// Tokenization (configuredVerifyArgv, the same allowlist the single-VerifyCmd path already
// uses) happens for ALL commands up front, before any subprocess is launched: since the
// verdict is AND, a single disallowed command already dooms the whole check, so there is no
// reason to pay for concurrent subprocesses whose result cannot change the outcome.
func tryParallelConfiguredVerify(targetDir string, commands []string, featureId int) automatedVerifyResult {
	argvs := make([][]string, len(commands))
	for i, raw := range commands {
		argv := configuredVerifyArgv(raw)
		if len(argv) == 0 {
			return failedVerify(fmt.Sprintf(
				"FAIL: verify command #%d is empty or uses disallowed shell operators: %s", i+1, raw))
		}
		argvs[i] = argv
	}

	// One goroutine per command, same idiom as runVerifyScript's own timeout race and
	// TaskRegistry.runWithTimeout: goroutine + buffered channel, no third-party library.
	// The channel is sized to the command count so no goroutine blocks trying to send.
	outcomes := make(chan parallelVerifyOutcome, len(commands))
	for i, argv := range argvs {
		go func(index int, argv []string, raw string) {
			outcomes <- runOneConfiguredVerify(targetDir, argv, raw, featureId, index)
		}(i, argv, commands[i])
	}

	ordered := make([]parallelVerifyOutcome, len(commands))
	for range argvs {
		o := <-outcomes
		ordered[o.index] = o
	}

	lines := make([]string, len(ordered))
	failures := 0
	for i, o := range ordered {
		lines[i] = o.line
		if !o.passed {
			failures++
		}
	}
	joined := strings.Join(lines, " | ")

	if failures == 0 {
		return passedVerify(fmt.Sprintf("PASS: all %d verify commands passed. %s", len(commands), joined))
	}
	return failedVerify(fmt.Sprintf("FAIL: %d of %d verify commands did not pass. %s", failures, len(commands), joined))
}

// runOneConfiguredVerify runs a single already-tokenized verify command (index is 0-based;
// display uses the 1-based n) and formats its line for the aggregate message, reusing the
// same suffix helpers (logSuffix, verifyOutputSuffix, firstMeaningfulLine/snippet) the
// single-command path already relies on.
func runOneConfiguredVerify(targetDir string, argv []string, raw string, featureId, index int) parallelVerifyOutcome {
	n := index + 1
	result := runVerifyScript(targetDir, argv)
	logPath := writeVerifyLogIndexed(targetDir, raw, featureId, n, result)

	if result.TimedOut {
		return parallelVerifyOutcome{
			index: index,
			line:  fmt.Sprintf("#%d TIMEOUT (%s): %s%s", n, verifyTimeoutDescription(), raw, logSuffix(logPath)),
		}
	}
	if result.ExitCode == 0 {
		return parallelVerifyOutcome{
			index:  index,
			passed: true,
			line:   fmt.Sprintf("#%d PASS: %s%s", n, raw, logSuffix(logPath)),
		}
	}
	return parallelVerifyOutcome{
		index: index,
		line:  fmt.Sprintf("#%d FAIL (exit %d): %s%s", n, result.ExitCode, raw, verifyOutputSuffix(result, logPath)),
	}
}

func exitCodeOf(err error) int {
	if err == nil {
		return 0
	}
	if exitErr, ok := err.(interface{ ExitCode() int }); ok {
		return exitErr.ExitCode()
	}
	return -1
}

func verifyTimeoutMs() int {
	timeoutMs := engine.CurrentConfig().TimeoutMs
	if timeoutMs <= 0 {
		return 0
	}
	margin := min(500, max(1, timeoutMs/10))
	return max(1, timeoutMs-margin)
}

func verifyTimeoutDescription() string {
	timeoutMs := verifyTimeoutMs()
	if timeoutMs <= 0 {
		return "no limit"
	}
	return fmt.Sprintf("%dms", timeoutMs)
}

// writeVerifyLog writes the single-command verify log — unchanged behavior, now a thin
// wrapper over writeVerifyLogIndexed with no check index (Go has no default parameters).
func writeVerifyLog(targetDir, command string, featureId int, result verifyScriptResult) string {
	return writeVerifyLogIndexed(targetDir, command, featureId, 0, result)
}

// writeVerifyLogIndexed writes one command's verify log. checkIndex is 1-based; 0 means
// "the only command" and keeps the original, un-suffixed path
// (.harness/logs/verify-feature-{featureId}.log) so single-VerifyCmd behavior and any
// external tooling that reads that path are untouched. A positive checkIndex (one of several
// concurrently run VerifyCmds) gets its own file, .harness/logs/verify-feature-{featureId}-{n}.log,
// so concurrent commands never race on the same log.
func writeVerifyLogIndexed(targetDir, command string, featureId, checkIndex int, result verifyScriptResult) string {
	fileName := fmt.Sprintf("verify-feature-%d.log", featureId)
	if checkIndex > 0 {
		fileName = fmt.Sprintf("verify-feature-%d-%d.log", featureId, checkIndex)
	}
	relativePath := filepath.Join(".harness", "logs", fileName)
	displayPath := strings.ReplaceAll(relativePath, "\\", "/")

	fullPath := relativePath
	if err := os.MkdirAll(filepath.Dir(fullPath), 0o755); err != nil {
		return fmt.Sprintf("log unavailable (%s)", oneLine(err.Error(), ""))
	}

	content := fmt.Sprintf(
		"timestampUtc: %s\ncommand: %s\ncwd: %s\nexitCode: %d\ntimedOut: %t\n\n--- stdout ---\n%s\n\n--- stderr ---\n%s",
		time.Now().UTC().Format("2006-01-02T15:04:05.0000000Z"), command, targetDir,
		result.ExitCode, result.TimedOut, result.Output, result.Error)

	if err := os.WriteFile(fullPath, []byte(content), 0o644); err != nil {
		return fmt.Sprintf("log unavailable (%s)", oneLine(err.Error(), ""))
	}

	return displayPath
}

func passResult(featureId int, output, errOutput, logPath string) string {
	firstLine := firstMeaningfulLine(output, errOutput)
	var result string
	if strings.HasPrefix(strings.ToUpper(firstLine), "PASS") {
		result = snippet(firstLine, 240)
	} else {
		result = fmt.Sprintf("PASS: verify-feature.sh %d passed", featureId)
	}
	return result + logSuffix(logPath)
}

func verifyOutputSuffix(result verifyScriptResult, logPath string) string {
	output := snippet(firstMeaningfulLine(result.Output, result.Error), 240)
	if strings.TrimSpace(output) == "" {
		return logSuffix(logPath)
	}
	return fmt.Sprintf(": %s%s", output, logSuffix(logPath))
}

func firstMeaningfulLine(values ...string) string {
	for _, value := range values {
		normalized := strings.ReplaceAll(value, "\r", "\n")
		for _, line := range strings.Split(normalized, "\n") {
			line = strings.TrimSpace(line)
			if line != "" {
				return line
			}
		}
	}
	return ""
}

func logSuffix(logPath string) string {
	if strings.TrimSpace(logPath) == "" {
		return ""
	}
	return fmt.Sprintf(". Log: %s", logPath)
}

func snippet(value string, maxChars int) string {
	text := oneLine(value, "")
	if len(text) <= maxChars {
		return text
	}
	return strings.TrimRight(truncateUtf8Bytes(text, maxChars), " \t") + "..."
}

func fileExistsLocal(path string) bool {
	info, err := os.Stat(path)
	return err == nil && !info.IsDir()
}

func configuredVerifyArgv(raw string) []string {
	text := strings.TrimSpace(raw)
	if text == "" || strings.ContainsAny(text, ";&|<>`$") {
		return nil
	}
	args := strings.Fields(text)
	if len(args) == 0 {
		return nil
	}
	bin := strings.ToLower(filepath.Base(args[0]))
	if bin == "sh" || bin == "bash" || bin == "zsh" || bin == "fish" || bin == "cmd" || bin == "powershell" || bin == "pwsh" {
		for _, arg := range args[1:] {
			if arg == "-c" || arg == "-command" || arg == "/c" {
				return nil
			}
		}
	}
	return args
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
